Pr3 add join leave UI buttons

haydnli-shopify opened the pull request:
# Summary

Implements join/leave project functionality with an intuitive UI integrated directly into the project members table. This PR builds on the member management endpoints from PR2 to provide a seamless user experience for project collaboration.

Builds on PR2 (#2779) and Issue(#2742) - requires member management endpoints

# Changes Made

## Frontend Implementation
- **Integrated join/leave UI** into existing `ProjectMembers` component
- **Added permission-based button states**: 
  - "Join Project" for non-members of public projects
  - "Leave Project" for regular members
  - "Owner Cannot Leave" (disabled) for project owners
- **Added navigation logic** to redirect users to project list after leaving projects
- **Removed redundant "Public Access" section** from project Settings page

## Translation & Localization
- Added translation keys for join/leave operations and success/error messages

# User Experience

## Join Project Flow
1. User views a public project they're not a member of
2. "Join Project" button appears in the members table
3. Click triggers API call and user becomes project member

## Leave Project Flow  
1. User views a project they're a member of (but not owner)
2. "Leave Project" button appears in members table
3. Click triggers API call and user is redirected to project list

## Permission Handling
- **Project owners** cannot leave (disabled button)
- **Regular members** can leave projects
- **Non-members** can join public projects only

haydnli-shopify committed:
fix: retrack PR2 based on PR1

haydnli-shopify committed:
fix: lint errors in test_projects.py

haydnli-shopify committed:
fix: update test assertions to include is_public field

haydnli-shopify committed:
style: apply pre-commit formatting fixes

haydnli-shopify committed:
Add project visibility update API endpoint

haydnli-shopify committed:
Fix user autocomplete: allow all users to see user list

haydnli-shopify committed:
Fix project deletion permissions: allow project owners to delete their projects

haydnli-shopify committed:
Apply pre-commit formatting fixes to PR2

haydnli-shopify committed:
Fix project deletion permissions and member race conditions

haydnli-shopify committed:
Add join/leave project UI functionality

haydnli-shopify committed:
Add join/leave project UI functionality

haydnli-shopify committed:
Allow public project access to gateways endpoints

haydnli-shopify committed:
Add Join/Leave button to project header and restrict CLI section to members

haydnli-shopify committed:
Add Join/Leave button to project header and restrict CLI section to members

r4victor commented:
Please address https://github.com/dstackai/dstack/pull/2759#issuecomment-2969369967

haydnli-shopify commented:
> Please address [#2759 (comment)](https://github.com/dstackai/dstack/pull/2759#issuecomment-2969369967)

Thanks for the feedback!
I’ve tested the UI as a non-member on public projects, and the page works as expected—member-only sections aren’t shown, and the UI doesn’t break.
Let me know if you spot any specific issues, but I believe the current logic in #2795 already handles this case well.

r4victor commented:
@haydnli-shopify, great, we can test it once the PR is ready for review.

haydnli-shopify commented:
> @haydnli-shopify, great, we can test it once the PR is ready for review.

Both ticket are ready for review
PR for backend APIs: https://github.com/dstackai/dstack/pull/2779
PR for UI: https://github.com/dstackai/dstack/pull/2795

haydnli-shopify committed:
fix: address PR comments from r4victor

olgenn commented on frontend/src/pages/Project/Form/index.tsx:
```diff
@@ -16,6 +16,7 @@ export const ProjectForm: React.FC<IProps> = ({ initialValues, onCancel, loading
 
     const formMethods = useForm<IProject>({
         defaultValues: {
+            is_public: false,
```
use camelCase

olgenn commented on frontend/src/pages/Project/Details/Settings/index.tsx:
```diff
@@ -113,14 +126,39 @@ export const ProjectSettings: React.FC = () => {
 
     const debouncedMembersHandler = useCallback(debounce(changeMembersHandler, 1000), []);
 
+    const changeVisibilityHandler = (is_public: boolean) => {
```
```suggestion
    const changeVisibilityHandler = (isPublic: boolean) => {
```
use camelCase, pls

olgenn commented on frontend/src/pages/Project/Details/index.tsx:
```diff
@@ -1,14 +1,131 @@
+        return getProjectRoleByUserName(project, userData.username);
+    }, [project, userData?.username]);
+
+    const isProjectOwner = useMemo(() => {
```
```suggestion
    const isProjectOwner = userData?.username === project?.owner.username;
```
Don't need to use memoization here

olgenn commented on frontend/src/pages/Project/Members/index.tsx:
```diff
@@ -30,6 +38,67 @@ export const ProjectMembers: React.FC<IProps> = ({ members, loading, onChange, r
+        return member?.project_role || null;
+    }, [members, userData?.username]);
+
+    const isProjectOwner = useMemo(() => {
```
```suggestion
    const isProjectOwner = userData?.username === project?.owner.username;
```
Don't need to use memoization here

olgenn commented on frontend/src/pages/Project/Members/index.tsx:
```diff
@@ -30,6 +38,67 @@ export const ProjectMembers: React.FC<IProps> = ({ members, loading, onChange, r
+    const isMember = currentUserRole !== null;
+    const isMemberActionLoading = isAdding || isRemoving;
+
+    const handleJoinProject = async () => {
```
I think handleJoin Project, handle Leave Project need to move to react hook(e. g. use Project Member Actions), it possible to remove duplicate code

haydnli-shopify committed:
refactor: improve code style and extract member actions hook

haydnli-shopify committed:
refactor: improve code style and extract member actions hook

olgenn commented on frontend/src/pages/Project/Details/index.tsx:
```diff
@@ -1,14 +1,129 @@
+    const isMember = currentUserRole !== null;
+    const isMemberActionLoading = isAdding || isRemoving;
+
+    const handleJoinProject = async () => {
```
Can we use handlers from `useProjectMemberActions` here?

haydnli-shopify committed:
refactor: use useProjectMemberActions in ProjectDetails

haydnli-shopify committed:
refactor: use useProjectMemberActions in ProjectDetails

peterschmidt85 commented:
Running on the state of 0.19.15:

```
  File "/Users/cheptsov/dstack/.venv/lib/python3.13/site-packages/alembic/script/base.py", line 286, in _catch_revision_errors
    raise util.CommandError(resolution) from re
alembic.util.exc.CommandError: Can't locate revision identified by '35e90e1b0d3e'
```

I assume, the PR branch should merge the changes from the master

peterschmidt85 commented on frontend/src/api.ts:
```diff
@@ -56,6 +56,9 @@ export const API = {
         SET_MEMBERS: (name: IProject['project_name']) => `${API.PROJECTS.DETAILS(name)}/set_members`,
+        ADD_MEMBERS: (name: IProject['project_name']) => `${API.PROJECTS.DETAILS(name)}/add_members`,
+        REMOVE_MEMBERS: (name: IProject['project_name']) => `${API.PROJECTS.DETAILS(name)}/remove_members`,
+        UPDATE_VISIBILITY: (name: IProject['project_name']) => `${API.PROJECTS.DETAILS(name)}/update_visibility`,
```
I'd recommend to have `/update` instead of `/update_visibility` and reuse this API for other updates that we may want to support in the feature.

peterschmidt85 commented:
(Minor) I'd suggest to move the UI for managing project visibility to the Danger Zone

peterschmidt85 reviewed (changes requested):
- [ ] Ensure you merge the changes from the master (and do a quick test) before merging the PR
- [ ] Replace `/update_visibility` with more generic `/update`
- [ ] Move the UI for managing the project to `Danger Zone`. Ensure to use a ComboBox (Public|Private) and ask for a confirmation.
- [ ] (Optional) Consider showing `Visibility` in the `Projects` table as a separate column

peterschmidt85 commented:
Plus, showing `Owners Cannot LEave` doesn't make sense.
<img width="1705" alt="Screenshot 2025-06-20 at 11 19 35" src="https://github.com/user-attachments/assets/e40a5b3a-c2ae-488b-b8e1-44f66be69f04" />

peterschmidt85 commented:
Also, the `Make project public` shows the state incorrectly
<img width="989" alt="Screenshot 2025-06-20 at 11 21 27" src="https://github.com/user-attachments/assets/0ee2e48b-9257-47c0-adb4-8b6d1d19407a" />

haydnli-shopify committed:
feat: generalize project update endpoint

haydnli-shopify committed:
refactor: rename /update_visibility to /update for more generic project updates

haydnli-shopify committed:
Fix linter errors and remove obsolete updateProjectVisibility mutation

haydnli-shopify committed:
Move project visibility to Danger Zone with ComboBox and add visibility column

haydnli-shopify committed:
Fix: Hide 'Leave Project' button for project owners

haydnli-shopify committed:
Hide Leave Project button for last admin instead of showing error

haydnli-shopify committed:
Fix project visibility display issue

peterschmidt85 reviewed (approved):
Added a few minor changes. Please review.
If all is good, please confirm so we can merge.

haydnli-shopify committed:
Remove redundant comments that duplicate code logic

haydnli-shopify committed:
Remove only redundant comments added in PR2

r4victor commented on src/dstack/_internal/server/routers/projects.py:
```diff
@@ -92,3 +97,60 @@ async def set_project_members(
+async def update_project(
+    body: UpdateProjectRequest,
+    session: AsyncSession = Depends(get_session),
+    user_project: Tuple[UserModel, ProjectModel] = Depends(ProjectManager()),
```
Project manager can manage project members but otherwise they are regular users, so they should not be able to update projects. Project update must be limited to project admins.

peterschmidt85 committed:
Minor changes:
 - UI captions
 - Added confirmation on Leave
 - Removed Join/Leave from Project Details Header
